### DISCUSSION TITLE
Can I use "Allow specific actors to bypass pull request requirements." to reduce PR noise?

### DISCUSSION THREAD

**Discussion by manuelmeister** [renovatebot/renovate#15123] [Request Help] [answered]
There is a new functionality called [Allow bypassing required pull requests
](https://github.blog/changelog/2021-11-19-allow-bypassing-required-pull-requests/) in the pull request rule of the branch protection rules.
![Highlighted new option "Allow bypassing required pull requests"](https://user-images.githubusercontent.com/3001985/163546372-e0ad7bbe-0540-4cc6-a652-0cf17842137f.png)

(How) Can this be used with the renovate bot / app, so It does not create "automerging" PRs with the two approve bots, if not necessary?

**Comment by viceice** [comment, marked as answer]
I think you need to enable that checkbox and add renovate username(s) to the then visible exclusion list. 🤷‍♂️

**Reply by manuelmeister** [reply]
I checked that, but when I type in "renovate" nothing appears.

**Reply by viceice** [reply]
@JamieMagee do you know something about it?

**Reply by JamieMagee** [reply]
Renovate is an App, not a user. It's a different permissions model. I suspect there is an equivalent permission that needs to be requested by @rarkins. But updating the Renovate app's permissions will trigger an email to all users asking them to grant the new permission.

**Reply by rarkins** [reply]
Permissions are also not optional either, meaning any new install has no option other then to accept it

**Reply by JamieMagee** [reply]
And until a user accepts, you'll get tokens with the old set of permissions. In practice this means that you can get a token with the set of permissions that your app had configured at any point in the past, and you have to gracefully handle this 😬

**Reply by viceice** [reply]
Did i misunderstood something? 🤔

I thougth @manuelmeister just want to add renovatebot to the bypass list, so renovate can automerge with branch.

Why does renovate need new permissins for that? 🤔

**Reply by JamieMagee** [reply]
Sorry, I'm not sure if it does require a new permission or not. I was explaining why Renovate didn't show up in the list.

I've not configured branch policies with Renovate before.

**Reply by rarkins** [reply]
I had assumed no new permissions are necessary for this but perhaps GitHub's config does not allow apps to bypass

**Reply by manuelmeister** [reply]
This would solve the need for the two approve bots, as the updates could just be automerged from the branch if CI passes.

**Reply by rarkins** [reply]
Note how "apps" is included here:
<img width="745" alt="image" src="https://user-images.githubusercontent.com/6311784/163666497-1ee9da95-3b86-41a6-81fa-aeefba6a34b9.png">

But not here:
<img width="467" alt="image" src="https://user-images.githubusercontent.com/6311784/163666513-e2708798-6ec6-48e5-8525-8dd6386907db.png">

**Reply by wKovacs64** [reply]
I assume adding Renovate to the "Restrict who can push to matching branches" list will **not** circumvent "Require a pull request before merging" + "Require approvals", eh?

**Reply by viceice** [reply]
you should ask this at GitHub forums, as it's nothing renovate can influence.

**Reply by manuelmeister** [reply]
> I assume adding Renovate to the "Restrict who can push to matching branches" list will **not** circumvent "Require a pull request before merging" + "Require approvals", eh?

Nope, that's why I was excited to see the "bypass specific actors".

**Reply by wKovacs64** [reply]
> > I assume adding Renovate to the "Restrict who can push to matching branches" list will **not** circumvent "Require a pull request before merging" + "Require approvals", eh?
> 
> Nope, that's why I was excited to see the "bypass specific actors".

Figured. Bummer. 😕

**Reply by rarkins** [reply]
Are you aware of Renovate Approve Bot?

**Reply by wKovacs64** [reply]
> Are you aware of Renovate Approve Bot?

I was not - that looks promising. In its description, it says "if the PR is configured for automerging." Does that mean if Renovate is configured to automerge it or does it require this 👇 option be enabled?

![image](https://user-images.githubusercontent.com/1288694/164753523-dfda2199-d110-4c79-bf24-17708767c7fa.png)

Edit: I'm testing this right now, just waiting for Renovate to run.

**Reply by rarkins** [reply]
Essentially it looks for "Automerge: enabled" in the PR body content and approves only if it sees that (otherwise it would just add noise)

**Reply by rarkins** [reply]
FYI it only approved new PRs after installation so won't apply to any existing PRs you have - rename and close them to generate a replacement.

**Reply by wKovacs64** [reply]
I can't tell if it worked yet, no bot activity and the Renovate dashboard seems down?

**Reply by rarkins** [reply]
Whoops, I missed the domain renewal today. Now renewed but it takes a while for the CNAME entries to refresh. https://renovate-app.netlify.app/ should work in the meantime

**Reply by wKovacs64** [reply]
It works! This solves our use-case, thanks Rhys! And I now realize the OP knew all this and doesn't want the PRs at all, which is slightly different. Sorry for the thread-jacking. 😬

**Reply by manuelmeister** [reply]
Exactly, I would like to have pull requests only when the CI fails. 😊

[My question on Github Community](https://github.community/t/allow-specific-actors-to-bypass-pull-request-requirements-with-apps-like-renovate/246515)

**Reply by wKovacs64** [reply]
https://github.blog/changelog/2022-05-17-consistently-allow-github-apps-as-exceptions-to-branch-protection-rules/

This works now, FYI! 🎉

**Reply by manuelmeister** [reply]
> This works now, FYI! 🎉

@viceice Does this mean that if I now enable this option and change nothing else, then the approval bots will no longer be needed?

**Reply by viceice** [reply]
i think yes

**Reply by wKovacs64** [reply]
> > This works now, FYI! 🎉
> 
> @viceice Does this mean that if I now enable this option and change nothing else, then the approval bots will no longer be needed?

Yes, I tried it and can confirm it works that way.

**Reply by manuelmeister** [reply]
Somehow it still creates pull requests, even though no check failed: https://github.com/ecamp/ecamp3/pull/2751

Am I doing it wrong?

**Reply by rarkins** [reply]
You need to look at the debug logs at the time of PR creation. Maybe Renovate is still detecting branch protection

**Reply by manuelmeister** [reply]
<details>
<summary>Log excerpt before pr creation</summary>

```bash
INFO: Branch created(branch="renovate/cypress-included-9.x")
{
  "commitSha": "13371c4f018e9cb8b9bbd7abbbba9bda263a334b"
}
DEBUG: Ensuring PR(branch="renovate/cypress-included-9.x")
DEBUG: There are 0 errors and 0 warnings(branch="renovate/cypress-included-9.x")
DEBUG: getBranchPr(renovate/cypress-included-9.x)(branch="renovate/cypress-included-9.x")
DEBUG: findPr(renovate/cypress-included-9.x, undefined, open)(branch="renovate/cypress-included-9.x")
DEBUG: findPr(renovate/cypress-included-9.x, undefined, closed)(branch="renovate/cypress-included-9.x")
DEBUG: resolveBranchStatus(branchName=renovate/cypress-included-9.x, ignoreTests=false)(branch="renovate/cypress-included-9.x")
DEBUG: getBranchStatus(renovate/cypress-included-9.x)(branch="renovate/cypress-included-9.x")
DEBUG: branch status check result(branch="renovate/cypress-included-9.x")
{
  "state": "pending",
  "statuses": []
}
DEBUG: No check runs found(branch="renovate/cypress-included-9.x")
{
  "result": {
    "total_count": 0,
    "check_runs": []
  }
}
DEBUG: Branch status yellow(branch="renovate/cypress-included-9.x")
DEBUG: Creating PR(branch="renovate/cypress-included-9.x")
{
  "prTitle": "chore(deps): update dependency cypress/included to v9.7.0"
}
DEBUG: Creating PR(branch="renovate/cypress-included-9.x")
{
  "title": "chore(deps): update dependency cypress/included to v9.7.0",
  "head": "ecamp:renovate/cypress-included-9.x",
  "base": "devel",
  "draft": false
}
DEBUG: PR created(branch="renovate/cypress-included-9.x")
{
  "pr": 2751,
  "draft": false
}
```


</details>

**Reply by manuelmeister** [reply]
It does not say anything about branch detection for this issue. However it says something about another:
```bash
DEBUG: Found PR #2722(branch="renovate/major-js-linter")
DEBUG: Found branch protection(branch="renovate/major-js-linter")
DEBUG: Branch protection: PR Reviews are required before merging(branch="renovate/major-js-linter")
```

**Reply by rarkins** [reply]
Please create a feature request on this. Renovate proactively detects branch protection to avoid trying a pointless automerge. I guess now it can't detect that it's allowed to bypass the branch protection, so we either need to detect its own bypass capabilities *or* stop doing proactive checks.

**Reply by BacLuc** [reply]
We successfully configured renovate to auto merge without PR into our repository.
More info: https://github.com/renovatebot/renovate/issues/16037#issuecomment-1163574524
